### PR TITLE
Fixed #3658

### LINKED ISSUE #3658: Latest Fable fails to compile Fable.React
### Description

```.\src\Client\fable_modules\Fable.React.9.3.0\Fable.React.Props.fs(43,5): (43,12) error FABLE: Cannot test erased union cases```

This is because the bundled FCS contains https://github.com/dotnet/fsharp/pull/16341, and Fable tries to create case tester functions for every case.

```js
export function HTMLAttr__get_IsHref(this$, unitArg) {
    if (this$.tag === 0) {
        return true;
    }
    else {
        return false;
    }
}

export function HTMLAttr__get_IsCustom(this$, unitArg) {
    if (this$.tag === 1) {
        return true;
    }
    else {
        return false;
    }
}

export const x = HTMLAttr__get_IsCustom(new HTMLAttr(1, []));
```

for

```fsharp
type HTMLAttr =
    | Href of string
    | Custom

let x = Custom.IsCustom
```

Adding `[<Fable.Core.Erase>]` to `Href` or `Custom` introduces a compiler error.

I suggest Fable is adjusted to recognize the newly exposed case testers and to emit tag comparisons at call sites instead of creating properties. That way compilation will only fail when trying to use a case tester on an erased case.

```js
export const x = (new HTMLAttr(1, [])).tag === 1;
```

### Repro code

https://fable.io/repl/#?code=C4TwDgpgBAEgKgWQDIEFjAE5QLwCgoFQA+UAFAFRQDaAPAGICGARgDYQB0AwgPYYcCiGBgGcIAPgC6UcgEoYfAGZRuS4ZgCWAOwDm+QiU4BXNdwC2uXG2BQAHjihGTp9gElhj4GaA&html=Q&css=Q

### Related information

* Fable version: 4.8.1

### PR DESCRIPTION
**Pull request by ncave**
- Inline union case testers (#3658)

### CODE REVIEW

**Comment by ncave** [comment]
@MangelMaxime I still think we should only suppress the testers for the erased union cases, and let bundlers eliminate the unused code (same as with other generated methods like reflection etc.).
I can make that change if you agree.

**Comment by MangelMaxime** [comment]
> I still think we should only suppress the testers for the erased union cases, and let bundlers eliminate the unused code (same as with other already generated methods like reflection etc.).

@ncave I am ok with doing it this way I just have a suggestion regarding the new warning (see comment above).

**Comment by ncave** [comment]
@MangelMaxime I updated it to inline all the union case testers instead.

File: src/Fable.Transforms/FSharp2Fable.fs [hunk 1]
```diff
@@ -1905,6 +1919,12 @@ let private transformMemberDecl
         []
     elif memb.IsImplicitConstructor then
         transformPrimaryConstructor com ctx memb args body
+    // Ignore union case testers for erased union cases
+    elif isErasedUnionCaseTester memb then
+        $"Erased union case tester will be ignored: {memb.LogicalName}"
+        |> addWarning com [] None
+
+        []
```
[the remark this answers is not in the document]
Comment (ncave) [reply src/Fable.Transforms/FSharp2Fable.fs]: Union case testers are now inlined, so it will do that.

**Comment by MangelMaxime** [comment]
Awesome thank you @ncave 

> P.S. I still think it's not a good idea to model free-form open-ended POJOs with F# unions, there must be a better way we can suggest for new bindings, but I guess for now this should fix existing legacy bindings.

I do agree on that, we have better ways of doing it now days using how Feliz works.

```fs
open Fable.Core
open Fable.Core.JsInterop

// Create an interface to force the representation
type IHTMlAttr = interface end

module Interop = 
    
    let inline mkIHTMlAttr (key : string) (value : obj) = unbox<IHTMlAttr> (key, value)

// Use an erased interface with inlined static members to represent the different cases
// The [<Erase>] attribute is used to not generate the constructor or reference information
// Inlining the static members minimizes the bundle size and also allows Fable 
// to generate a Literal POJO later on
// 
// Benefit of using an interface with static member versus a module and functions
// is that we gain access to overload making it possible to avoid U2<string, float>
// and instead use two separate overloads
[<Erase>]
type HTMLAttr =
    
    static member inline Href (value : string) =
        Interop.mkIHTMlAttr "href" value

    static member inline Custom (key : string) (value : string) =
        unbox<IHTMlAttr> (key, value)

// Create an helper function to guide the user in creating the object
// This is inlined to, to give a chance to Fable to generate a Literal POJO
let inline createHtmlAttr (a : IHTMlAttr list) = 
    createObj (unbox a)

// Here a literal POJO is generated
let myInstance =
    createHtmlAttr [ 
        HTMLAttr.Href "http://www.google.com" 
        HTMLAttr.Custom "data-delay" "100ms"
    ]
```

```js
export const myInstance = {
    href: "http://www.google.com",
    "data-delay": "100ms",
};
```

I should probably add similar exemple to Fable documentation officially explaining this trick to people and when documenting the `keyValueList` mark it has obsolete compared to this code.

Thinking about it, it is perhaps possible to upgrade Fable.React to this new style while keeping the double list syntax. The difficulty however would be how to support SSR via this style. If someone think it is worth it and want to give it a try and think it feel free to send a PR.

**Pull request merged**